proposals/0000-dps-compact.md

```diff
@@ -0,0 +1,393 @@
+
+```c
+stg_compactAddHollowzh(
+  P_ compact, // the compact region where the constructor heap object will be added
+  W_ info     // the info table pointer that will be set as header for the new constructor heap object
+) {
+  [...]
+  // find address where to store the pointer to the new constructor object
+  pp = compact + SIZEOF_StgHeader + OFFSET_StgCompactNFData_result;
+  // reuse existing macro to allocate inside the compact region
+  ALLOCATE(compact, size, p, new_ctor_obj, tag);
+  // write the supplied info table pointer
+  SET_HDR(new_ctor_obj, info, CCS_SYSTEM);
+  // store the pointer to the new constructor in the return location
+  P_[pp] = new_ctor_obj;
+  [...]
+
+  return (P_[pp], SIZEOF_StgHeader, SIZEOF_W);
+}
+```
+
+`compactAddHollow#` allocates space for a new constructor heap object in the compact region, but doesn't set any of its fields (so it is _hollow_). It only sets its header with the supplied info table pointer. The user is responsible for setting the fields of the constructor heap object before reading it. The primitive returns the newly allocated constructor heap object, but also the size of the header of the object, and the size of the words on the platform so that the user can later set the fields of the returned object.
+
+### Retrieving the info table pointer: primtype `InfoTablePtrOf#` and compile-time primop `reifyInfoTablePtr#`
+
+`compactAddHollow#` is in charge of allocating the constructor heap object in the compact region, but it doesn't have access to the info table pointer for the constructor. All information about info tables has long been inlined/erased at runtime. So we need a primitive `reifyInfoTablePtr#` to reify the info table pointer of a given constructor at compile-time into a static value that the RTS can then use.
+
+```haskell
+type InfoTablePtrOf# :: forall k. k -> TYPE AddrRep
+
+reifyInfoTablePtr# :: forall k (a :: k). (# #) -> InfoTablePtrOf# a
+```
+
+Primtype `InfoTablePtrOf#` shares the same representation as `Addr#` so it serves both as a sort of input at type level to select the data constructor that `reifyInfoTablePtr#` should operate on, and as an output at value level to carry the resulting info table pointer.
+
+We pass the desired data constructor as a type to `reifyInfoTablePtr#`, using type ascription for the return value:
+
+```haskell
+let leftInfoPtr :: InfoTablePtrOf# 'Left = reifyInfoTablePtr# (# #)
+```
+
+`reifyInfoTablePtr#` is a compile-time primops (i.e. substituted with actual Cmm code in the _Stg to Cmm_ phase). It inspects the type of its return value (which is accessible in the code generation pipeline for compile-time primops `emitPrimOp`):
+
+- if it is of shape `InfoTablePtrOf# 〈s〉`, where `〈s〉` is a `Symbol`, it returns the info table pointer for the Stg object `stg_〈s〉`
+- if it is of shape `InfoTablePtrOf# '〈d〉`, where `〈d〉` is a data constructor, it returns the info table pointer for data constructor `〈d〉`
```

Comment:
I guess `reifyInfoTablePtr#` doesn't “return” since this all happens at compile time. It is, rather, “compiled into”, isn't it?

Reply:
Changed.